src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -193,38 +228,42 @@ DispatchResult<V> dispatch() {
         DataLoaderInstrumentationContext<DispatchResult<?>> instrCtx = ctxOrNoopCtx(instrumentation().beginDispatch(dataLoader));
 
         boolean batchingEnabled = loaderOptions.batchingEnabled();
-        final List<K> keys;
-        final List<Object> callContexts;
-        final List<CompletableFuture<V>> queuedFutures;
-        try {
-            lock.lock();
 
-            int queueSize = loaderQueue.size();
-            if (queueSize == 0) {
-                lastDispatchTime.set(now());
-                instrCtx.onDispatched();
-                return endDispatchCtx(instrCtx, emptyDispatchResult());
+        LoaderQueueEntry<K, V> loaderQueueEntryHead;
+        while (true) {
+            loaderQueueEntryHead = loaderQueue.get();
+            if (loaderQueue.compareAndSet(loaderQueueEntryHead, null)) {
+                break;
             }
-
-            // we copy the pre-loaded set of futures ready for dispatch
-            keys = new ArrayList<>(queueSize);
-            callContexts = new ArrayList<>(queueSize);
-            queuedFutures = new ArrayList<>(queueSize);
-
-            loaderQueue.forEach(entry -> {
-                keys.add(entry.getKey());
-                queuedFutures.add(entry.getValue());
-                callContexts.add(entry.getCallContext());
-            });
-            loaderQueue.clear();
+        }
+        if (loaderQueueEntryHead == null) {
             lastDispatchTime.set(now());
-        } finally {
-            lock.unlock();
+            instrCtx.onDispatched();
+            return endDispatchCtx(instrCtx, emptyDispatchResult());
+        }
+        int queueSize = calcQueueDepth(loaderQueueEntryHead);
+        // we copy the pre-loaded set of futures ready for dispatch
+        Object[] keysArray = new Object[queueSize];
+        CompletableFuture[] queuedFuturesArray = new CompletableFuture[queueSize];
+        Object[] callContextsArray = new Object[queueSize];
+        int index = queueSize - 1;
+        while (loaderQueueEntryHead != null) {
```

Comment:
You dont have t use an array and then asList it.  Create the ArrayList and the use `java.util.ArrayList#set` to set a specific index

The ArrayList is an array under the covers - so tis the same.  Actually the performance cost is the same as what you have here but the codes simpler

Reply:
Also you dont need to calculate queue size if you use ArrayList - but then again its defaults to size 10 - so maybe growing the array is more expensive over time then walking the list twice.  We could keep a list size in a voilaitle int say